### PR TITLE
Register devices on join

### LINKED ISSUE #425: Register new devices on join
This is a **feature request** for **the Backend**.

Currently, only devices that are registered can be accepted when joining the network. This requires all devices to be known, including their DevEUI. This might not be practical in large scale deployments.

Therefore, it should be possible for devices to register on join. To avoid join challenges to be sent to all Handlers that registered the AppEUI with a Broker, there should be a first-come-first-serve registration of AppEUI. We might leverage the current unique registration of AppEUI/DevEUI but then with a blank DevEUI. If the Broker doesn't find a specific AppEUI/DevEUI combination for the join, it can challenge a Handler that has the AppEUI registered with a DevEUI as catchall.

## On Security

When allowing any DevEUI to join for an AppEUI and the AppKey leaks, we basically enable spoofing of existing devices and registering new devices. This means that the application needs its own mechanism to identify the device. Many IoT platforms already support this with a device registry. Also, when introducing this feature, we should inform the application owners of the potential risk.

Also, we should restrict the AppEUIs that can register devices anonymously to a range in the Foundation's EUI block. Otherwise, people can abuse The Things Network to send conflicting join accepts for any join request as AppEUIs are public. We can configure the Brokers with an environment variable for a range.

### PR DESCRIPTION
**Pull request by htdvisser**
This PR adds functionality for registering devices when they join the network.

The steps to do this are as follows:

1. `ttnctl applications select [AppID] [AppEUI]`
2. `ttnctl devices register on-join [Device ID Prefix] [AppKey]`
    - This will add an access key to your application if needed
3. Join a device with AppEUI and AppKey - Any DevEUI is accepted
4. The device is registered, its DevID will be `Device ID Prefix`-`DevEUI`

Closes #425

### CODE REVIEW

File: core/handler/activation.go [hunk 1]
```diff
@@ -236,15 +249,54 @@ func (h *handler) HandleActivation(activation *pb_broker.DeduplicatedDeviceActiv
...
 	}
 
 	return res, nil
 }
+
+func (h *handler) registerDeviceOnJoin(base *device.Device, activation *pb_broker.DeduplicatedDeviceActivationRequest) (*device.Device, error) {
+	clone := base.Clone()
+	clone.DevID = strings.ToLower(fmt.Sprintf("%s-%s", base.DevID, activation.DevEui.String()))
+	clone.DevEUI = *activation.DevEui
+	clone.Description = fmt.Sprintf("On-Join device registered on %s", time.Now().UTC().Format(time.RFC3339))
```
Comment (johanstokking) [inline core/handler/activation.go]: "Registered on join on %s"

Also can we have a more human readable date/time format here? It doesn't have to be parsable; we have created and updated time stamps

File: core/handler/activation.go [hunk 2]
```diff
@@ -236,15 +249,54 @@ func (h *handler) HandleActivation(activation *pb_broker.DeduplicatedDeviceActiv
...
+
+	app, err := h.applications.Get(base.AppID)
+	if err != nil {
+		return nil, errors.Wrap(err, "Application not registered to this Handler")
+	}
+
+	if app.RegisterOnJoinAccessKey == "" {
+		return nil, errors.NewErrInvalidArgument("Application", "Does not have RegisterOnJoinAccessKey")
```
Comment (johanstokking) [inline core/handler/activation.go]: Does this error propagate to the user? If so, make it human friendly

File: core/handler/activation_test.go [hunk 3]
```diff
@@ -69,105 +78,182 @@ func TestHandleActivation(t *testing.T) {
...
-	var wg WaitGroup
-
-	appEUI := types.AppEUI{1, 2, 3, 4, 5, 6, 7, 8}
-	appID := appEUI.String()
-	devEUI := types.DevEUI{1, 2, 3, 4, 5, 6, 7, 8}
-	devID := devEUI.String()
-	unknownDevEUI := types.DevEUI{8, 7, 6, 5, 4, 3, 2, 1}
 
-	appKey := [16]byte{1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16}
+	// var wg WaitGroup
```
Comment (johanstokking) [inline core/handler/activation_test.go]: What's this?

File: ttnctl/cmd/devices_register_on_join.go [hunk 4]
```diff
@@ -0,0 +1,126 @@
...
+	"github.com/TheThingsNetwork/ttn/api/protocol/lorawan"
+	"github.com/TheThingsNetwork/ttn/core/types"
+	"github.com/TheThingsNetwork/ttn/ttnctl/util"
+	"github.com/spf13/cobra"
+)
+
+const registerOnJoinAccessKeyName = "register-on-join"
+
+var devicesRegisterOnJoinCmd = &cobra.Command{
+	Use:   "on-join [Device ID] [AppKey]",
```
Comment (johanstokking) [inline ttnctl/cmd/devices_register_on_join.go]: Make clear that this is a prefix. Also, be consistent with casing; `DeviceID` or `DevID` (whatever is used elsewhere)

**Comment by johanstokking** [comment]
Looping in @dotpy3 for sanity checking

File: ttnctl/cmd/devices_register_on_join.go (hunk 4 above)
Comment (htdvisser) [inline ttnctl/cmd/devices_register_on_join.go]: Will change this to "Device ID Prefix". Taking care of consistency across ttnctl should be done in a separate "refactor" PR

**Comment by dotpy3** [comment]
Overall LGTM, but shouldn't we offer a command in `ttnctl` to disable on-join registration, such as `ttnctl on-join off`? I don't know if it will be clear for users that on-join registration is disabled by removing the template device. Or, we could add an additional info line when enabling on-join registration, such as "Registered on-join device template - remove this device to disable on-join registration". Do the users need to know how on-join registration works within TTN?

**Comment by htdvisser** [comment]
I think the documentation would be a better place to explain this. People who use _and_ ttnctl _and_ on-join registration already know what they're doing.

**Pull request merged**